### ISSUE TITLE
Wrong URL match

### ISSUE THREAD

**Issue by piegamesde** [keepassxreboot/keepassxc-browser#656]
I just got a wrong URL match using the KeepassXC browser extension.

URL in question: `https://sso.tu-darmstadt.de/login`
URLs of matched entries: `http://https//aur.archlinux.org https://archlinux.org`, `https://passport.twitch.tv https://www.twitch.tv/`
Expected, correct match (snip): `https://sso.tu-darmstadt.de/login […] https://print.informatik.tu-darmstadt.de/user`

If I click on "deny", the popup closes and the correct credentials are filled in anyway.

## Steps to Reproduce (for bugs)
I cannot reproduce this deterministically yet (you can try creating entries with the urls above), but this is not the first time this happens. When it happens, it is always with the same two false positives.

## Debug info

KeePassXC - 2.5.0
KeePassXC-Browser - 1.5.3
Operating system: Arch Linux
Browser: Chromium 78.0.3904.70

**Comment by droidmonkey** [comment]
Do you have multiple urls stuffed into the URL field of the (an) entry?

**Comment by piegamesde** [comment]
@droidmonkey Yes. But this does not justify such matching behaviour in any way.

**Comment by droidmonkey** [comment]
Use the new multi-url feature in the entry. Reduce the url field to the primary, then go to the browser integration page of the entry editor, add additional URLs.

By stuffing urls into the url field it confuses the URL parser which outputs junk.

**Comment by varjolintu** [comment]
Please note that the URL field is not meant for multiple URL's.

**Comment by piegamesde** [comment]
If there is an option to be conservative in URL matching and that option is enabled, there is no excuse for false positives, no matter how bullshit the input is. If you make assumptions about the contents of the URL field, either enforce it in the input field or deal with it correctly.

**Comment by droidmonkey** [comment]
Not disagreeing with you there

**Comment by piegamesde** [comment]
So, I changed all my entries with multiple URLs and it does not recognize any of them except the "main" URL. But I don't know if this is another bug.

**Comment by droidmonkey** [comment]
Be careful with http vs https.

**Comment by piegamesde** [comment]
@droidmonkey I have https everywhere, enforce protocol matching and don't use www.

@varjolintu In the original issue, I had the URLs as simple space separated list in the field.

**Comment by varjolintu** [comment]
@piegamesde It's this function: https://github.com/keepassxreboot/keepassxc/blob/develop/src/browser/BrowserService.cpp#L996

The parsing is currently done with `QUrl::TolerantMode`. ` QUrl::StrictMode` would be much better, and using that with `isValid()` would probably make things right.

**Comment by LeBaux** [comment]
> Use the new multi-url feature in the entry. Reduce the url field to the primary, then go to the browser integration page of the entry editor, add additional URLs.

> By stuffing urls into the url field it confuses the URL parser which outputs junk. We opt to match liberally by default, you can turn on strict matching in the settings.

Not to be rude, but shouldn't the parser be able to deal with this? I probably have some deformed entry **somewhere** in my database and because of it, nothing works. One of my databases has over 1000 entries, a lot of them in Recycle bin. So now I should audit all of it by hand? Feels like the password manager should, you know, manage it. At least detect broken URL fields?

I understand you probably have a reason to change how the parser works, but if it breaks on malformed URL it does not feel like it does it job well. This was introduced with KeepassXC 2.5.0, it is possible the issue is tied to main program but I found this opened issue first so I am just adding my 2c.

![image](https://user-images.githubusercontent.com/5657511/68012022-2bf97a80-fc89-11e9-93c4-810f36c55136.png)

Thank you.

**Comment by libklein** [comment]
I propose not matching invalid urls (i.e. by modifying https://github.com/keepassxreboot/keepassxc/blob/develop/src/browser/BrowserService.cpp#L996 accordingly). 

I will prepare a pull request in https://github.com/keepassxreboot/keepassxc/. Should we perhaps move the issue there? Or at least open a new one and reference it from this issue?

**Comment by droidmonkey** [comment]
There is already an issue on the main application board

**Comment by varjolintu** [comment]
@LeBaux We are going to fix this soon.

**Comment by piegamesde** [comment]
Is there any good reason to do some fuzzy matching on URLs anymore? It feels to me that liberal matching was more like a workaround because we only had one URL per entry for a long time. Is there any disadvantage of strict matching that cannot be worked around by adding more specific URLs to an entry?

**Comment by droidmonkey** [comment]
@varjolintu can you move this to the KeePassXC board?
